test/end2end_test.go

```diff
@@ -2844,7 +2845,8 @@ func testStreamsQuotaRecovery(t *testing.T, e env) {
 			defer wg.Done()
 			payload, err := newPayload(testpb.PayloadType_COMPRESSABLE, 314)
 			if err != nil {
-				t.Fatal(err)
+				t.Error(err)
```

Comment:
I think a te.cancel() here is necessary to avoid a hang in the main goroutine.  You could test this by changing it to `if err == nil` above.